### PR TITLE
feat: Add support for smoothly resuming training from a saved checkpoint

### PR DESCRIPTION
**Pull request by Abhishek-TAMU**
### Description of the change

In this PR, have added feature of resuming the tuning from the last saved checkpoint of last tuning using the flag 
`--output_dir` which stores all the checkpoints.

### Related issue number

#[1007](https://github.ibm.com/ai-foundation/watson-fm-stack-tracker/issues/1007)

### How to verify the PR

Resuming the tuning of model after initial tuning and verify the output checkpoints. 

### Was the PR tested

- [ ] I have added >=1 unit test(s) for every new method I have added.
- [ ] I have ensured all unit tests pass

### CODE REVIEW

File: tuning/utils/config_utils.py
Comment (fabianlim) [inline tuning/utils/config_utils.py:140]: why is this logic needed? since it seems to be [duplicated from that inside HF trainer](https://github.com/huggingface/transformers/blob/8820fe8b8c4b9da94cf1e4761876f85c562e0efe/src/transformers/trainer.py#L1930)

Comment (anhuong) [reply tuning/utils/config_utils.py:140]: This is also doing similar work to what exists in [build/utils](https://github.com/foundation-model-stack/fms-hf-tuning/blob/main/build/utils.py#L43-L63) so that or this method could be refactored to be used in both places

**Review by kmehant — CHANGES_REQUESTED** [review]
How is this different from simply passing the flag https://github.com/huggingface/transformers/blob/8820fe8b8c4b9da94cf1e4761876f85c562e0efe/src/transformers/training_args.py#L663?

**Comment by ashokponkumar** [comment]
Why not just set the [`resume_from_checkpoint`](https://github.com/huggingface/transformers/blob/984bc11b0882ff1e5b34ba717ea357e069ceced9/src/transformers/trainer.py#L1858) field to `True` and let HF trainer take care of it?

File: tuning/utils/config_utils.py
Comment (Abhishek-TAMU) [reply tuning/utils/config_utils.py]: That's correct. I missed that one. Now importing it from `transformers.trainer_utils import get_last_checkpoint`

Comment (kmehant) [reply tuning/utils/config_utils.py]: I think it would be good to reuse this flag to pass to train() function as is the way user provides, this would save us from too much of custom handling which limit other potential usecases.

https://github.com/huggingface/transformers/blob/8820fe8b8c4b9da94cf1e4761876f85c562e0efe/src/transformers/training_args.py#L663

Comment (Abhishek-TAMU) [reply tuning/utils/config_utils.py]: Just to get clarity @kmehant, does user passes the flag `resume_from_checkpoint` if users wants to resume from particular checkpoint ?
And in particular case where user doesn't pass the flag, do we assume that user doesn't intend on resuming the tuning and wants to start a new one, OR we still do need to check for latest checkpoint path and use it to resume tuning. 
Thanks!

File: tuning/sft_trainer.py
Comment (kmehant) [inline tuning/sft_trainer.py:368]: `resume_from_checkpoint=True` gives same effect isn't it?

File: tuning/utils/config_utils.py
Comment (kmehant) [reply tuning/utils/config_utils.py]: > Just to get clarity @kmehant, does user passes the flag resume_from_checkpoint if users wants to resume from particular checkpoint ?

Yes, `--resume_from_checkpoint=true` if to start from last checkpoint or to a path if it has to be some arbitrary checkpoint. And then we pass this as is to the train()

> And in particular case where user doesn't pass the flag, do we assume that user doesn't intend on resuming the tuning and wants to start a new one, OR we still do need to check for latest checkpoint path and use it to resume tuning.

I have no particular suggestion on the default flow when user does not pass the flag 🙂

Comment (Abhishek-TAMU) [reply tuning/utils/config_utils.py]: >Yes, --resume_from_checkpoint=true if to start from last checkpoint or to a path if it has to be some arbitrary checkpoint. And then we pass this as is to the train()

Okay

File: tuning/sft_trainer.py
Comment (Abhishek-TAMU) [reply tuning/sft_trainer.py:368]: `resume_from_checkpoint=True`, Yes I tested it. It gives the same effect if the `output_dir` has checkpoints available or else throws error.

Comment (Abhishek-TAMU) [reply tuning/sft_trainer.py:368]: It would be helpful for me to know your requirements on default flow when user does not pass the flag as then could be able to finalize the code changes for review.

Comment (Abhishek-TAMU) [reply tuning/sft_trainer.py:368]: @ashokponkumar Any comment/input on above ?

Comment (ashokponkumar) [reply tuning/sft_trainer.py:368]: What error do we get when `resume_from_checkpoint` is set to True and there is no checkpoint? We can go with this fix for now, but it would be good to check in the HF Trainer, if they are okay with gracefully failing. Even if `resume_from_checkpoint` is set, if there is no checkpoint in the output dir, it should start like a normal run from scratch.

Comment (ashokponkumar) [reply tuning/sft_trainer.py:368]: Also, if user actually provides a false explicitly we should honor it.

Comment (Abhishek-TAMU) [reply tuning/sft_trainer.py:368]: >What error do we get when resume_from_checkpoint is set to True and there is no checkpoint

Error output and failing of code and exiting.
`ValueError: No valid checkpoint found in output directory (outputs/lora-tuning)`

Comment (Abhishek-TAMU) [reply tuning/sft_trainer.py:368]: >Even if resume_from_checkpoint is set, if there is no checkpoint in the output dir, it should start like a normal run from scratch.

>Also, if user actually provides a false explicitly we should honor it.

Okay.

Comment (Abhishek-TAMU) [reply tuning/sft_trainer.py]: Based on above comments, I have pushed few changes.
@kmehant Feel free to review it. 

Cases covered and tested:

1- **If user passes the flag `resume_from_checkpoint` with value** `False`: Tuning will start from scratch.

2- **If user passes the flag `resume_from_checkpoint` with value** `True`: if `output_dir` has checkpoints, then tuning will resume from last checkpoint, else tuning will start from scratch.

3- **If user passes the flag `resume_from_checkpoint` with value** `String: Entire Path of checkpoint-x`: The model would be loaded from `Path of checkpoint-x` and tuning will resume from the `checkpoint-x`.

4- **If user doesn't pass flag `resume_from_checkpoint`** : if `output_dir` has checkpoints, then tuning will resume from last checkpoint, else tuning will start from scratch.

**Comment by kmehant** [comment]
Thanks @Abhishek-TAMU 

@ashokponkumar FYA https://github.com/foundation-model-stack/fms-hf-tuning/pull/300#issuecomment-2302860601

**Comment by ashokponkumar** [comment]
Sounds good.

> Based on above comments, I have pushed few changes.
> @kmehant  Feel free to review it. 
> 
> Cases covered and tested:
> 
> 1- **If user passes the flag `resume_from_checkpoint` with value** `False`: Tuning will start from scratch.
> 
> 2- **If user passes the flag `resume_from_checkpoint` with value** `True`: if `output_dir` has checkpoints, then tuning will resume from last checkpoint, else tuning will start from scratch.
> 
> 3- **If user passes the flag `resume_from_checkpoint` with value** `String: Entire Path of checkpoint-x`: The model would be loaded from `Path of checkpoint-x` and tuning will resume from the `checkpoint-x`.
> 
> 4- **If user doesn't pass flag `resume_from_checkpoint`** : if `output_dir` has checkpoints, then tuning will resume from last checkpoint, else tuning will start from scratch.
> 
> 
> 
> 
>

**Comment by kmehant** [comment]
@Abhishek-TAMU can we add simple unit tests for each of the cases you mentioned above? Thanks

**Comment by Abhishek-TAMU** [comment]
>@Abhishek-TAMU can we add simple unit tests for each of the cases you mentioned above? Thanks

Sure! Have added the test cases and pushed the changes. Thanks!

File: tuning/sft_trainer.py
Comment (ashokponkumar) [inline tuning/sft_trainer.py:379]: Why not simplify this logic even further and move it into __post_init__ of training_args. 

We can check if `resume_from_checkpoint` is `None` or `True`, and if so, we check if `get_last_checkpoint(training_args.output_dir)` returns a non empty string. If it returns a non empty string, then we set `resume_from_checkpoint` to `True`, else we set `resume_from_checkpoint` to `False`.  In the current condition added, if `get_last_checkpoint` does not return a checkpoint, doesn't it get set to empty string, which is not what we want, isn't it?

We can leave every other case untouched, Isn't it? HF will already handle the other cases well, or is it not handling when `resume_from_checkpoint` is a string `false`?

**Review by ashokponkumar — COMMENTED** [review]
one query..

File: tuning/sft_trainer.py
Comment (Abhishek-TAMU) [reply tuning/sft_trainer.py:379]: >if `get_last_checkpoint` does not return a checkpoint, doesn't it get set to empty string, which is not what we want, isn't it?

`get_last_checkpoint` either returns a checkpoint string or None.

Comment (Abhishek-TAMU) [reply tuning/sft_trainer.py:379]: > is it not handling when `resume_from_checkpoint` is a string false ?

When `resume_from_checkpoint` is a string false, `trainer()` needs argument `resume_from_checkpoint` value as boolean.

`trainer()` accepts argument `resume_from_checkpoint` value as boolean True/False, None (none_type), or string with path of checkpoint.

Comment (ashokponkumar) [reply tuning/sft_trainer.py:379]: So we are adding support for string `true` and `false` additionally, even though HF Trainer does not support it?

Comment (ashokponkumar) [reply tuning/sft_trainer.py:379]: What is the behaviour if we set the Training argument `resume_from_checkpoint` to be `True` or `False`? Isn't it handling it as a boolean `True` or `False` and behave accordingly?

Comment (Abhishek-TAMU) [reply tuning/sft_trainer.py:379]: Training argument `resume_from_checkpoint` takes in the path to a folder with a valid checkpoint for the model, as mentioned [here](https://github.com/huggingface/transformers/blob/7591ca5bc583aadbdfe6b2fb36a12c874233323c/src/transformers/training_args.py#L665), and doesn't work with `boolean/String` value `True/False`

Comment (ashokponkumar) [reply tuning/sft_trainer.py:379]: Then lets go ahead with this change. LGTM. We can refine later if required.

**Review by ashokponkumar — DISMISSED** [review]

File: tuning/sft_trainer.py
Comment (anhuong) [inline tuning/sft_trainer.py:368]: Is this saying that if resume_from_checkpoint is not set, then it would enable resume from checkpoint? That seems off...If a user doesn't pass anything in, why would it be enabled?

File: tuning/sft_trainer.py
Comment (anhuong) [inline tuning/sft_trainer.py:375]: Am I understanding the logic here correctly -- that `resume_from_training` can be passed in as a boolean value or as a path to a directory to resume training from? Would be good to provide docs on this if the behavior is different from how HF transformers handles the argument as I see HF transformers only accepts path to checkpoint
> resume_from_checkpoint (str, optional) — The path to a folder with a valid checkpoint for your model.

Comment (Abhishek-TAMU) [reply tuning/sft_trainer.py:375]: Sure!

Mentioning `trainer.train()` [arguments](https://github.com/huggingface/transformers/blob/db70426854fe7850f2c5834d633aff637f14772e/src/transformers/trainer.py#L1901).

```
resume_from_checkpoint (`str` or `bool`, *optional*):
If a `str`, local path to a saved checkpoint as saved by a previous instance of [`Trainer`]. If a
`bool` and equals `True`, load the last checkpoint in *args.output_dir* as saved by a previous instance
of [`Trainer`]. If present, training will resume from the model/optimizer/scheduler states loaded here.
```

If the value of `resume_from_checkpoint` is passed `None` or `False` to `trainer.train()`, then it would behave similiar as not passing this flag to `trainer.train()`

File: tests/test_sft_trainer.py
Comment (anhuong) [inline tests/test_sft_trainer.py:210]: Try to make this one more succinct, Also what is `total_flos`?

Comment (anhuong) [reply tests/test_sft_trainer.py:210]: Let's make this a LoRA test to test resuming training from diff tuning techniques

**Review by anhuong — CHANGES_REQUESTED** [review]
Great changes and much appreciate the tests Abhishek! Have a few refactors that can be made to the tests to create cleaner code. Do you know how much time this adds to the tests by running tuning an additional 7 times?

File: tuning/sft_trainer.py
Comment (anhuong) [reply tuning/sft_trainer.py:375]: Ah I was looking at the docs here - https://huggingface.co/docs/transformers/main_classes/trainer#transformers.TrainingArguments.resume_from_checkpoint

File: tuning/sft_trainer.py
Comment (anhuong) [inline tuning/sft_trainer.py:369]: Since this does accept a boolean value, shouldn't this just check the bool value? A user isn't passing in a string "True" but the boolean "True" right?

File: tests/test_sft_trainer.py
Comment (anhuong) [reply tests/test_sft_trainer.py:262]: Can be a future improvement

File: tuning/sft_trainer.py
Comment (Abhishek-TAMU) [reply tuning/sft_trainer.py:368]: Yes, I can update the Readme. In this [readme](https://github.com/Abhishek-TAMU/fms-hf-tuning/blob/main/README.md), is there any specific section recommended, where I can mention above details for users (regarding info of `resume_from_checkpoint` flag) ?

File: tests/test_sft_trainer.py
Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py:125]: It's definitely a good test to add in general to test tuning, but not sure how it is related to resuming from checkpoint functionality.

File: tests/test_sft_trainer.py
Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py:109]: Yes! This is addressed!

File: tests/test_sft_trainer.py
Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py:240]: Added the description!

File: tests/test_sft_trainer.py
Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py:257]: Added the description!

File: tests/test_sft_trainer.py
Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py:267]: Yes this change is helpful. Thanks!
Also I guess the apt function name could be `_get_training_logs_by_epoch`. Do you agree with this ?

File: tests/test_sft_trainer.py
Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py:262]: Good point. Can look into this one as a future improvement.

File: tests/test_sft_trainer.py
Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py:264]: Thank you for reviewing test cases. This is done!

File: tests/test_sft_trainer.py
Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py:118]: Yes I checked it and the first one would always be epoch 1. And taking in your refactor, I can just modify it to something like below: (as I was taking out the loss in the code, but now just taking log history)

```
init_log_history = init_trainer_state["log_history"][0]
assert init_log_history["epoch"] == 1

final_log_history = final_trainer_state["log_history"][0]
assert final_log_history["epoch"] == 1

assert init_log_history["loss"] == final_log_history["loss"]
```

File: tests/test_sft_trainer.py
Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py:87]: Make sense. Thanks!

File: tests/test_sft_trainer.py
Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py:250]: In line 248-250, we don't necessarily take out last checkpoint, instead get checkpoint based on `checkpoint_index`, so I guess `transformers.utils.get_last_checkpoint()` may not be useful here.

File: tests/test_sft_trainer.py
Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py:136]: Sure! Have removed the description.

File: tests/test_sft_trainer.py
Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py:177]: Okay got it. Thanks!

File: tests/test_sft_trainer.py
Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py:195]: Yes correct, that's the logic. Also comment seems misleading, hence changed it to:

`# Training again with higher epoch and same output dir`

File: tests/test_sft_trainer.py
Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py:210]: `total_flos` is basically Total Floating Point Operations (FLOPs).
Formal def:
`total_flos` is a measure of the computational workload performed by the model during training. It is computed as the cumulative sum of FLOPs (Floating Point Operations) for each training step.

Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py:210]: > Let's make this a LoRA test to test resuming training from diff tuning techniques

Sure! Made this test as LoRA test. Thanks!

**Comment by Abhishek-TAMU** [comment]
> Great changes and much appreciate the tests Abhishek! Have a few refactors that can be made to the tests to create cleaner code. Do you know how much time this adds to the tests by running tuning an additional 7 times?

Thank you @anhuong for the review on test cases. I have made changes mentioned and pushed it. Also I calculated the time these 4 test cases adds to the tests. Difference in running with and without the 4 test case is 15 seconds. Thanks!

File: tuning/sft_trainer.py
Comment (Abhishek-TAMU) [reply tuning/sft_trainer.py:376]: Yes, `resume_from_checkpoint` does accepts bool in `trainer.train()`. Its just that, when user passes `resume_from_checkpoint` flag as bool, it gets taken as string as it gets treated as flag in [training arguments](https://huggingface.co/docs/transformers/main_classes/trainer#transformers.TrainingArguments.resume_from_checkpoint) which just accepts string.
image

File: tests/test_sft_trainer.py
Comment (anhuong) [reply tests/test_sft_trainer.py]: i could go either way on this, its only getting by epoch if the user provides an epoch, otherwise it returns the full training logs. so its not necessarily by epoch but is fine either way to me

File: tests/test_sft_trainer.py
Comment (anhuong) [inline tests/test_sft_trainer.py:230]: nit: kinda repetitive so makes more succinct
```suggestion
    Get the trainer state from the latest or specified checkpoint directory.
    The trainer state is returned along with the path to the checkpoint.
```

File: tests/test_sft_trainer.py
Comment (anhuong) [reply tests/test_sft_trainer.py]: true

File: tests/test_sft_trainer.py
Comment (anhuong) [reply tests/test_sft_trainer.py]: Not the description, I was referring to the test. All tests should have descriptions but this test looks to be doing the same work as the one above

File: tests/test_sft_trainer.py
Comment (anhuong) [inline tests/test_sft_trainer.py:193]: nit: Could be a little clearer and should update that this is for LoRA
```suggestion
    Test resume checkpoint from a specified checkpoint path for LoRA tuning.
```

File: tests/test_sft_trainer.py
Comment (anhuong) [inline tests/test_sft_trainer.py:190]: SHould include LoRA here
```suggestion
def test_resume_training_from_checkpoint_with_flag_checkpoint_path_lora():
```

File: tuning/sft_trainer.py
Comment (anhuong) [reply tuning/sft_trainer.py:376]: ahh thanks for the clarity

**Review by anhuong — COMMENTED** [review]
Not sure why all of my comments got split up in different reviews, sorry about that. There are a few nits on the descriptions and small changes that can be addressed as well as a test case that I'm not sure if its useful. In addition, a quick note should be added to docs that explains how this works but point at the HF.TrainingArguments / Trainer docs

**Comment by anhuong** [comment]
In terms of the linting and too many lines in test_sft_trainer.py, I think its okay to increase this value in pylint. We can create an issue to look into the tests and see how we can split them up into separate files. For now, lets increase to 1200

File: tests/test_sft_trainer.py
Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py]: Okay got it. Yes this test also checks if tuning is resuming from checkpoint but here I am marking `resume_from_checkpoint` flag as true, as to test the logic in the `sft_trainer` when flag value is True.

File: tests/test_sft_trainer.py
Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py:263]: Thank you!

File: tests/test_sft_trainer.py
Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py:230]: Thank you!

File: tests/test_sft_trainer.py
Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py:193]: Thank you!

**Comment by Abhishek-TAMU** [comment]
Have pushed changes related to function description, adding ReadMe and increase in line value in pylint. 
Feel free to review once again. @anhuong 
Thanks!

File: tests/test_sft_trainer.py
Comment (anhuong) [inline tests/test_sft_trainer.py:124]: I'm not sure if this method is useful or if it;s doing the same as the one above.

**Review by anhuong — APPROVED** [review]
Other than question on test, change looks good. Please answer question but i'll merge this as-is

**Comment by anhuong** [comment]
@kmehant Since you reviewed this PR and requested changes, the PR is waiting for your approval before it can be merged

File: tests/test_sft_trainer.py
Comment (Abhishek-TAMU) [reply tests/test_sft_trainer.py:124]: Yes, this test also checks if tuning is resuming from checkpoint but here I am marking `resume_from_checkpoint` flag as true, as to test the logic in the `sft_trainer.py` when flag value is True, though it's behavior would be similar to when value of `resume_from_checkpoint` is `None`.

Let me know, if you think we should remove this test.

**Review by kmehant — APPROVED** [review]
@Abhishek-TAMU lets rebase thank you.

File: tests/test_sft_trainer.py
Comment (anhuong) [reply tests/test_sft_trainer.py:124]: It's a bit redundant as resume_from_checkpoint=None creates the same behavior as resume_from_checkpoint=True but fine to leave for now

**Pull request merged**
